pkg/netdevice/vdpa.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2020 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package netdevice
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+
+	vdpa "github.com/redhat-virtio-net/govdpa/pkg/kvdpa"
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/types"
+)
+
+const (
+	vdpaTypeVirtio = "virtio"
+	vdpaTypeVhost  = "vhost"
+)
+
+/*vdpaTypeToDriver translates vdpaTypes (as specified in the netDevice selectors)
+  to vdpa bus drivers*/
+var vdpaTypeToDriver = map[string]string{
+	vdpaTypeVirtio: vdpa.VirtioVdpaDriver,
+	vdpaTypeVhost:  vdpa.VhostVdpaDriver,
+}
+
+type vdpaInfo struct {
+	device vdpa.VdpaDevice
+}
+
+// GetVdpaDevice returns a VdpaDevice
+func GetVdpaDevice(pciAddr string) vdpa.VdpaDevice {
+	vdpaDev, err := vdpa.GetVdpaDeviceByPci(pciAddr)
+	if err != nil {
+		glog.Infof("%s - No vDPA device found: %s", pciAddr, err)
+		return nil
+	}
+	return vdpaDev
+}
+
+// vdpaInfoProvider is a DeviceInfoProvider that handles the API information of vdpa-capable devices
+type vdpaInfoProvider struct {
+	dev      vdpa.VdpaDevice
+	vdpaType string
+}
+
+// NewVdpaInfoProvider returns a new InfoProvider associated with the given VDPAInfo
+func NewVdpaInfoProvider(vdpaType string, vdpaDev vdpa.VdpaDevice) types.DeviceInfoProvider {
+	vdpaInfoProvider := &vdpaInfoProvider{
+		vdpaType: vdpaType,
+		dev:      vdpaDev,
+	}
+	if err := vdpaInfoProvider.Error(); err != nil {
```

Comment:
I followed the same logic as with Rdma devices.
- At creation time, if the configuration in the ResourceConfig does not match the physical information, do not fail and let the filtering (selectors) do it's job of not adding those devices to the pool.
- Check again at runtime (i.e: GetDeviceSpecs) and, if it fails "again", then error.

When I looked at this I thought that this logic could be improved but I did not want to mix it with this PR. Now that the resourceConfig is available at pciNetDevice creation time, some of the filtering functionality can be moved there, e.g: do not create a pci device if `ResourceConfig.IsRdma != rdmaSpec.IsRdma` or, in this case, if `ResourceConfig.vdpaType != vpdDevice.GetDriver()`.
If we also degrade [this error log](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/blob/61fe3e908bff9314cbfc83078943a38144162b86/pkg/netdevice/netDeviceProvider.go#L52) to a warning, this would yield the extra checks in `netDeviceProvider::GetFilteredDevices` useless.

WDYT?

Reply:
I see, thanks for clarifying.

>When I looked at this I thought that this logic could be improved but I did not want to mix it with this PR. Now that the resourceConfig is available at pciNetDevice creation time, some of the filtering functionality can be moved there, e.g: do not create a pci device if ResourceConfig.IsRdma != rdmaSpec.IsRdma or, in this case, if ResourceConfig.vdpaType != vpdDevice.GetDriver().
If we also degrade this error log to a warning, this would yield the extra checks in netDeviceProvider::GetFilteredDevices useless.

This can be discussed in a separate PR, in my mind the "Device" should just expose what it has (passing in resource config is a way to tell the `New` method which extra attributes it should attempt to detect/expose).

Filtering should occur  as a separate step.